Replace library/asn1.php & ASNValue with phpseclib

nupplaphil opened the pull request:
- Replace `library/asn1.php` & removed a lot of unneeded Crypto methods
- Replace `library/ASNValue.class.php`

Therefore I chose a common security library (http://phpseclib.sourceforge.net).

A optional next part could be a rewrite of `Crypto.php` with using all common functions of phpseclib, so we get rid of our custom-scripted security features. AFAIK in terms of security, it's better to rely on well tested standard libraries.

And after the PR, we're more ready for using `Composer 2.0` (see all the composer notices after an install).

nupplaphil committed:
Replace library/asn1.php with phpseclib

MrPetovan reviewed (commented):
Have you verified it works the same?

nupplaphil commented:
That's why it's a draft :-)
Next step is to add tests for both methods to proof it.

nupplaphil commented:
even better :-) the phpseclib allows us to remove custom security code because of builtin functionality.

I added a diaspora public rsa key, extracted the needed values with the "old" functionality, saved it to files and replaced it with the new version => worked :)

MrPetovan commented:
Worked, huh?

nupplaphil committed:
Replace library/ASNValue.class.php with phpseclib functions

MrPetovan commented on src/Util/Crypto.php:
```diff
@@ -65,183 +65,53 @@ public static function rsaVerify($data, $sig, $key, $alg = 'sha256')
-		return $key;
+		$rsa = new RSA();
+		$rsa->loadKey([
+								  'e' => new BigInteger($e, 256),
```
Indentation is wrong here, it should be just 3 tabs.

MrPetovan commented on src/Util/Crypto.php:
```diff
@@ -65,183 +65,53 @@ public static function rsaVerify($data, $sig, $key, $alg = 'sha256')
-		unset($lines[0]);
-		unset($lines[count($lines)]);
-		$x = base64_decode(implode('', $lines));
+		$publicKey = new RSA();
```
Please rename `$publicKey` to `$RSA`.

MrPetovan commented on src/Util/Crypto.php:
```diff
@@ -65,183 +65,53 @@ public static function rsaVerify($data, $sig, $key, $alg = 'sha256')
-		unset($lines[0]);
-		unset($lines[count($lines)]);
-		$x = base64_decode(implode('', $lines));
+		$publicKey = new RSA();
```
Please rename `$publicKey` to `$RSA`.

MrPetovan commented on tests/src/Util/CryptoTest.php:
```diff
@@ -51,6 +53,62 @@ public function testRandomDigitsRandomInt()
+
+		$expectedRSA = new RSA();
+		$expectedRSA->loadKey([
+								  'e' => new BigInteger($e, 256),
```
Indentation should be 3 tabs here.

nupplaphil committed:
Format/name changes

MrPetovan commented on src/Util/Crypto.php:
```diff
@@ -73,10 +73,11 @@ public static function rsaVerify($data, $sig, $key, $alg = 'sha256')
-								  'e' => new BigInteger($e, 256),
-								  'n' => new BigInteger($m, 256)
-							  ]);
+		$rsa->loadKey(
+			[
+				'e' => new BigInteger($e, 256),
+				'n' => new BigInteger($m, 256)
+			]);
```
Please choose either the extended form
```suggestion
		$rsa->loadKey(
			[
				'e' => new BigInteger($e, 256),
				'n' => new BigInteger($m, 256)
			]
		);
```
Or the condensed form
```suggestion
		$rsa->loadKey([
			'e' => new BigInteger($e, 256),
			'n' => new BigInteger($m, 256)
		]);
```
But not a mix between both.

Same for the other instance.

nupplaphil replied:
hrmpf .. I had to reinstall my IDE .. Seems like there're some settings left to do ;-)

nupplaphil committed:
Fix indentation .. again ;-)

MrPetovan commented:
One less file in `library`!

MrPetovan commented:
Related to #4116
